documentation/high-level-introduction.md

```diff
@@ -0,0 +1,251 @@
+     
+    <Target Name="_CodesignNativeLibraries">
+                    <CodesignNativeLibraries
+                            SessionId="$(BuildSessionId)"
+                            AppBundleDir="$(AppBundleDir)"
+                            >
+                    </CodesignNativeLibraries>
+    </Target>
+```
+- You can attach a C# debugger to the task, but that can be tricky on mono since there is no attach to process
+- `SessionId` is needed on most C# tasks to enable “build remoting” for Visual Studio Windows.
+
+## Execution Model
+
+A very inaccurate but useful way of thinking about how msbuild processes is to split it into three parts.
+
+- Loading
+  - Starting with the csproj load each file included into a giant buffer
+    - Think C preprocessor gloming files together
+  - Some default values will be set in the global property dictionary
+- Property Eval
+  - Starting from the top, evaluate every node
+    - Property and Item Group will set items in the global context
+    - UsingTask will load task assemblies and prepare to invoke them
+    - Targets will define then and their ordering\inputs\output
+      - **Their contants will not be executed yet**
+  - This will respect conditions, and be done top to bottom based upon include order
+- Target Eval
+  - Starting with the defined “DefaultTargets", which for our purposes is always “Build” determine a list of target to invoke and their ordering
+  - Invoke one at a time, comparing their Input\Output against file timestamps like make
+
+
+## "SDK style” project and nuget injection
+  - SDK style projects and nugets are beyond the scope of this document 
+  - One important thing to note is the “magic” injection of files
+    - When referenced the props file will be “auto injected” to the top of your project and the targets injected at the very bottom. 
+    - Thus you can include files for these project types you don’t explicitly see referenced in the csproj
+
+
+## Debugging
+  - Easily the most important thing is to get a full diagnostic build. This will list all variables set and the ordering\execution of each target.
+  - You can add debug “Console.WriteLine”s by using the built in MessageTask:
+
+
+            <Message Text="Project File Name = $(MSBuildProjectFile)" />
+
+    - **Do note these must be inside targets**. They are only executed during Target Eval time.
+    - If you put them inside ItemGroups or PropertyGroups they will be ignored.
+  - There is a new “binary log” hotness which acts as a full diagnostics build but is faster and contains even more information. You can enable it by passing `/bl`  to msbuild. Read more about it [here](https://github.com/Microsoft/msbuild/blob/master/documentation/wiki/Binary-Log.md). 
+    - The support for loading .binlog files in Visual Studio for Mac is still somewhat new
+  - If you are developing on Windows you can add a `System.Diagnostics.Debugger.Break()` invocation and attach Visual Studio to the process.
```

Comment:
I tested this recently, and it seems you need `Debugger.Launch()` now.

Reply:
I think whether Break prompts to open a debugger is some system wide just-in-time debugging setting. I don't actually know how to turn it on.